### PR TITLE
doc: restore environment variables page

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
This was moved into the application development guide in commit
bf9ceac2ef82ddd8b4436d12016d70ffff849bbf
(" doc: move env_vars into application development").

That commit claimed the following rationale:

  No reason why we are duplicating this information in a standalone
  guide when we already talk about env variables in the application
  developer guide.

A github comment later says the duplication was in "Important build
system variables":

https://github.com/zephyrproject-rtos/zephyr/pull/34364#discussion_r616646517

This is incorrect and a misreading of the application development
guide.

The application development guide is talking about **cmake**
variables. These are not the same thing at all as **environment**
variables; treating them as if they are the same is simply wrong and
misleading.

In fact, the separate environment variables page was created for a
reason, which was to have a single place to explain to people unused
to managing their environment how things work, without cluttering
things up for people who already understand this.

There is nothing wrong with small, self-contained guides and in fact I
would argue they are a good thing.

Fix this confusing situation by restoring the original, separate page.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
Note that the rationale for a separate page is also explained in the original commit which added the page: 5edb6d5def80322759222d514418b84c00b8c7ed

**Review by gmarull — APPROVED** [review]

File: doc/guides/env_vars.rst
Comment (tejlmand) [inline doc/guides/env_vars.rst:109]: I think we should add a 5th option.

> Option 5: CMake pre-load script using `cmake -C <pre-load>.cmake`
>
> Chose this option if you have different sets of option you want to easily be able to choose between.
> This allows you to easily switch between a pre-defined set of options shared among multiple applications, for example:
> `cmake -Czephyr-sdk_release.conf` ... or `cmake -Cgnuarmemb_debug.cmake` which could then contain:
> ```
>  # zephyr-sdk_release.cmake
>  set(CONF_FILES=prj_release.conf CACHE STRING "Release conf file")
>  set(ZEPHYR_TOOLCHAIN_VARIANT zephyr CACHE STRING "Zephyr SDK")
> ```
> or 
> ```
> # gnuarmemb_debug.cmake
> set(CONF_FILES=prj_debug.conf CACHE STRING "Debug conf file")
> set(ZEPHYR_TOOLCHAIN_VARIANT gnuarmemb CACHE STRING "GNU arm embedded")
> ```

**Review by tejlmand — DISMISSED** [review]
With such a variety of options described, then I think we should inform the reader about `-C<pre-load>.cmake` as that is the only option that allows a set of different setups to easily choose from.

`.zephyrrc`, Zephyr build config package, `.bashrc` etc. kind of locks the system to one specific setup, whereas the `-C` actually allows an easy way to support a pre-defined set of options to choose from.

File: doc/guides/env_vars.rst
Comment (tejlmand) [reply doc/guides/env_vars.rst:109]: I just noticed, the title of this page specifically says:
https://github.com/zephyrproject-rtos/zephyr/blob/b10beefc7573b105f90a26dfa1c99633d3040f63/doc/guides/env_vars.rst#environment-variables
> Environment Variables

The 4th Option of using a `Zephyr` build configuration package is not setting any environment variables.

So the title in that case is misleading.

Also CMake variables can be set using `-D<variable>=<value>` which is not mentioned at all.
This could give an impression that the best way to set `CONF_FILES` would be through one of those ways.

Note, not all Zephyr CMake variables are being sourced from environment setting.
For example `ZEPHYR_MODULES` is not, but surprisingly someone added `ZEPHYR_EXTRA_MODULES` here: https://github.com/zephyrproject-rtos/zephyr/pull/19301.
Similar with `BOARD_ROOT` and `SOC_ROOT`.

File: doc/guides/env_vars.rst
Comment (tejlmand) [inline doc/guides/env_vars.rst:155]: So if we cannot create a comprehensive list or link to one, how will users of Zephyr ever be able to discover what settings that can be controlled using environment variables ?

Comment (mbolivar-nordic) [reply doc/guides/env_vars.rst:155]: > So if we cannot create a comprehensive list or link to one, how will users of Zephyr ever be able to discover what settings that can be controlled using environment variables ?

As usual, they will have to read the source code and ask questions from the developers. I hate this as much as you do, but it is important to at least document reality.

File: doc/guides/env_vars.rst
Comment (mbolivar-nordic) [reply doc/guides/env_vars.rst:109]: > The 4th Option of using a `Zephyr` build configuration package is not setting any environment variables.
> 
> So the title in that case is misleading.

Ugh, this is worse than I thought :slightly_frowning_face: 

If you read the PR description, you will see that this content was originally in a page that was all about environment variables. Then someone got confused and mixed up cmake and environment variables and put the environment variable page into the application development page near the cmake variable setting content.

Then apparently at some point after that, the two really got mixed up even more.

**Comment by mbolivar-nordic** [comment]
@tejlmand nice catch -- I've sent a v2 which hopefully fixes things by separating the cmake variable settings in option 4 from the environment variables in the previous sections.

**Comment by tejlmand** [comment]
> @tejlmand nice catch -- I've sent a v2 which hopefully fixes things by separating the cmake variable settings in option 4 from the environment variables in the previous sections.

I think this makes it much harder for the readers to understand the different ways they may set Zephyr CMake build variables.

I think it would be better to change this into setting of Zephyr build variables, and then describe that the Zephyr build system also supports setting the most important variables in the environment.

Afterall, the important thing here is how to set the Zephyr build variable, and not whether this is as environment or through another mean (build config package / cache pre-load).

File: doc/guides/env_vars.rst
Comment (tejlmand) [reply doc/guides/env_vars.rst]: > If you read the PR description, you will see that this content was originally in a page that was all about environment variables. Then someone got confused and mixed up cmake and environment variables and put the environment variable page into the application development page near the cmake variable setting content.

but maybe we should take a step back, and instead describe how to set Zephyr build variables as that is what is important, and not whether it's done as environment variables.

And those can be set with the principles you already had located together in the first version, including setting those as an environment variable.

File: doc/guides/env_vars.rst
Comment (tejlmand) [inline doc/guides/env_vars.rst:150]: just to make this even worse, then `DTC_OVERLAY_FILE` is not being picked up from environment.
It can only be set as a CMake variable (cache / local).

File: doc/guides/env_vars.rst
Comment (tejlmand) [reply doc/guides/env_vars.rst:143]: > > So if we cannot create a comprehensive list or link to one, how will users of Zephyr ever be able to discover what settings that can be controlled using environment variables ?
> 
> As usual, they will have to read the source code and ask questions from the developers. I hate this as much as you do, but it is important to at least document reality.

Ok, I created a complete list of all variables Zephyr supports as environment settings.
```
ZEPHYR_BASE
BOARD
SHIELD
CONF_FILE

WEST_DIR
ZEPHYR_EXTRA_MODULES
ZEPHYR_BOARD_ALIASES (Not supported as regular CMake variables, points to file, wrongly documented)

Toolchains:
TOOLCHAIN_ROOT
ZEPHYR_SDK_INSTALL_DIR
ZEPHYR_TOOLCHAIN_VARIANT (see list of possible values)
  - espressif
  - armclang
  - xtensa
  - llvm
  - arcmwdt
  - xtools
  - host
<toolchain_upper>_TOOLCHAIN_PATH

CROSS-COMPILE
ONEAPI_ROOT (alternative to ONEAPI_TOOLCHAIN_PATH)
XCC_USE_CLANG (Not supported as regular CMake variable, not documented)

CLANG_ROOT_DIR (deprecated, use LLVM_TOOLCHAIN_PATH)
GCCARMEMB_TOOLCHAIN_PATH (deprecated, use GNUARMEMB_TOOLCHAIN_PATH)


Emulators:
QEMU_BIN_PATH (has a CMake possibility of `QEMU=<path-to-qemu>`)
OVMF_FD_PATH (Not supported as regular CMake variables)
QEMU_EXTRA_FLAGS (not documented)
ARMFVP_BIN_PATH (has a CMake possibility of `ARMFVP=<path-to-armfvp>`)
ARMFVP_BL1_FILE
ARMFVP_FIP_FILE

Board specific env (Not supported as regular CMake variables)
EVERGLADES_SPI_GEN
EVERGLADES_SPI_CFG
MEC172X_SPI_GEN
MEC172X_SPI_CFG

Babble sim (Not supported as regular CMake variables):
BSIM_COMPONENTS_PATH
BSIM_OUT_PATH
```

I would say we should list the top-ones, and the important toolchain related:
`ZEPHYR_TOOLCHAIN_VARIANT`: (same descr as now)
`<toolchain_upper>_TOOLCHAIN_PATH: The path to the toolchain specified in `ZEPHYR_TOOLCHAIN_VARIANT`.
If `ZEPHYR_TOOLCHAIN_VARIANT=llvm` then this path is `LLVM_TOOLCHAIN_PATH`.

**Note**: `DTC_OVERLAY_FILE` is not being picked up from environment.

Comment (tejlmand) [reply doc/guides/env_vars.rst:143]: we can then refer to board and emulator environment variables in more generic manner.

Something like:
> Emulators and boards may depend on additional programs.
> The build system will try to locate those programs automatically, but may rely on additional CMake or environment variables to do so.
> Please consult the emulator and/or board specific documentation for more information.

That way we can provide a complete list of general build settings to the reader, while still point to the fact that emulators and boards may have more settings.

At least that would make me happier.

Comment (tejlmand) [reply doc/guides/env_vars.rst:143]: see also: https://github.com/zephyrproject-rtos/zephyr/issues/40389

File: doc/guides/env_vars.rst
Comment (mbolivar-nordic) [reply doc/guides/env_vars.rst:150]: >  then `DTC_OVERLAY_FILE` is not being picked up from environment.

Sorry, what do you mean by "then"? What is the "if"?

File: doc/guides/env_vars.rst
Comment (mbolivar-nordic) [reply doc/guides/env_vars.rst:143]: > Ok, I created a complete list of all variables Zephyr supports as environment settings.

OK, but can we deal with that in a separate PR? The scope of this PR is just to resurrect the environment variables page.

File: doc/guides/env_vars.rst
Comment (mbolivar-nordic) [reply doc/guides/env_vars.rst]: > how to set Zephyr build variables as that is what is important, and not whether it's done as environment variables.

I disagree; the point of the page was to explain how to set environment variables.

File: doc/guides/env_vars.rst
Comment (tejlmand) [reply doc/guides/env_vars.rst:150]: a stray `n`, should have been:
> the DTC_OVERLAY_FILE is not being picked up from environment.

File: doc/guides/env_vars.rst
Comment (tejlmand) [reply doc/guides/env_vars.rst]: I disagree there.

We should describe common ways to set Zephyr build variables together, that is:
- `-D<var>=value`
- Environment one time: `export <var>=<val>` / `set <var>=<val>`
- All terminals: `~/.bashrc` / `~/.bash_profile` / `rapidee`
- Zephyr rc: `~/.zephyrrc` / `zephyrrc.cmd`
- Zephyr build configuration package.
- CMake pre-load cache: `-C <preload>.cmake`

That way the user can decide which principle is best suited to their need.

Having just environment variables in a dedicated page but the Zephyr build configuration package as a note, and not mentioning the `-C` preload option in any way may lead the user to believe the environment settings are the best or recommended way to go.

I think we should describe the possibilities together and let the reader determine what is best option.
Not scattering the different possibilities in various pages, making it harder to get the full picture on this.

File: doc/guides/env_vars.rst
Comment (tejlmand) [reply doc/guides/env_vars.rst:143]: Accepted.

**Comment by mbolivar-nordic** [comment]
v5: 

- rebased to use the groups introduced in commit b48276e9954a11ccf1ebeb93c635f5f16fe8fdb2
- addressed @tejlmand reivew

**Comment by tejlmand** [comment]
>     * addressed @tejlmand review

except this: https://github.com/zephyrproject-rtos/zephyr/pull/40233#discussion_r753001604
> I think we should describe the possibilities together and let the reader determine what is best option.
Not scattering the different possibilities in various pages, making it harder to get the full picture on this.

The current page https://docs.zephyrproject.org/latest/application/index.html#setting-variables describes `Setting variables` not environment variables:
> Setting Variables
> Option 1: Just Once
> ...
> Option 2: In all Terminals

and https://docs.zephyrproject.org/latest/application/index.html#important-build-system-variables `Build system variables`, not environment.

> Important Build System Variables
> ...

So to be honest, I prefer the existing description over a dedicated environment page.
But a dedicated `settings` / `build system varibles` page would be the top.

Which is also what I mention here:
https://github.com/zephyrproject-rtos/zephyr/pull/40233#issuecomment-970033473

@nashif comments on this, you were the original author moving this together and cleaning up here: https://github.com/zephyrproject-rtos/zephyr/pull/34364
https://github.com/zephyrproject-rtos/zephyr/pull/34364#discussion_r616646517

I don't like thing just moving back and forth.

**Comment by mbolivar-nordic** [comment]
@tejlmand did you read the PR description? It explains why the current organization created by @nashif is wrong. In particular:

> This is incorrect and a misreading of the application development
> guide.

> The application development guide is talking about cmake
> variables. These are not the same thing at all as environment
> variables; treating them as if they are the same is simply wrong and
> misleading.

**Comment by mbolivar-nordic** [comment]
> @nashif comments on this ? you were the original author moving this together and cleaning up here: #34364 [#34364 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/34364#discussion_r616646517)
> 
> I don't like thing just moving back and forth.

... and I don't like my environment variable documentation getting mangled up like this in a PR that I wasn't asked for review on, only to get blocked when I try to fix it again.

**Comment by tejlmand** [comment]
> The application development guide is talking about cmake variables. 

Where ?
From what I read it talks about **Setting variables**[0] and **Build system variables**[1], and those can be controlled through environment variables or CMake (cache) variables for Zephyr.
[0]: https://docs.zephyrproject.org/latest/application/index.html#setting-variables
[1]: https://docs.zephyrproject.org/latest/application/index.html#important-build-system-variables

and we should keep that consistent, which I also raise here: https://github.com/zephyrproject-rtos/zephyr/issues/40389

I think we should keep the description of how you can set build system variables together, because the fact that Zephyr allows the same vars to be set both using CMake cache and environment is a bit special and I would like that documented together.

>  treating them as if they are the same is simply wrong and misleading.

and that's why we should describe that those particular build settings can be controlled either way, and with a clear description of which variable (environment vs CMake cache (vs CMake)) takes precedence.

> ... and I don't like my environment variable documentation getting mangled up like this in a PR that I wasn't asked for review on

I understand, but I was never part of #16064 and I happen to have an opinion on that too, which means I don't think it should simply be re-resurrected as it was but adjusted, as I said:
> but maybe we should take a step back, and instead describe how to set Zephyr build variables as that is what is important, and not whether it's done as environment variables. https://github.com/zephyrproject-rtos/zephyr/pull/40233#discussion_r750038288

> But a dedicated settings / build system varibles page would be the top. https://github.com/zephyrproject-rtos/zephyr/pull/40233#issuecomment-987763036

**Comment by tejlmand** [comment]
Dismissing my review.
I disagree we should have a special page on environment variables and split how users may configure the build system, as I have written here:
https://github.com/zephyrproject-rtos/zephyr/pull/40233#discussion_r753001604
https://github.com/zephyrproject-rtos/zephyr/pull/40233#issuecomment-970033473

Having the use of `-D<var>=<val>` and `set(<var> <val>)` in one place, the use of environment settings at another page, and finally the `build_config_package` at a third, and not mentioning a pre-load cache at all is messy to me.

Therefore I cannot give +1, but won't block this PR.
Someone else must approve.

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
I completely agree that this is clearer for the user than what we have now. That said, @tejlmand makes good points, especially when it comes to having a single unified location for documenting how the build can be configured in general. But I'd rather take small steps in that direction, and the proposal is better than the current state of things.

**Pull request merged**
